src/libraries/System.Text.Json/tests/System.Text.Json.Tests/Serialization/MetadataTests/DefaultJsonTypeInfoResolverTests.JsonPropertyInfo.cs

```diff
@@ -823,7 +823,8 @@ public static void DefaultIgnoreConditionFromOptionsDoesntFlowToShouldSerializeP
                 if (ti.Type == typeof(TestClassWithNumber))
                 {
                     Assert.Equal(1, ti.Properties.Count);
-                    Assert.Null(ti.Properties[0].ShouldSerialize);
+                    Assert.NotNull(ti.Properties[0].ShouldSerialize);
```

Comment:
I'd expect `ShouldSerialize = null` brings me back default behavior from options and not be equivalent to `always serialize`.

Reply:
So, we have two approaches to consider:

1. `ShouldSerialize = null` defaults back to the global `JsonIgnoreCondition` configuration. This is the existing approach.
2. The value of `ShouldSerialize` already reflects the global `JsonIgnoreCondition` configuration. If a user explicitly unsets the delegate, then it is equivalent `JsonIgnoreCondition.Never`. This is the approach introduced by this PR.

I think both approaches are valid, to an extent. The global setting is respected in both cases, so it simply becomes a question of when it gets consulted. I would tend to favor the latter approach, according to the principle that configuration on an individual type should override global configuration. That being said, the forcing function that brought about this change is the fact that post-hoc consultation of `JsonIgnoreCondition` is a source of inconsistency as illustrated by #71908.

Reply:
I was trying to compare what NSJ does in this case but they seem to have DefaultValueHandling both on options and property but FWIW null `DefaultValueHandling` on property means `use options default`. IMO that confirms me that `ShouldSerialize=null` should follow options defaults

Reply:
Couple of points:

1. If we take this approach for `DefaultIgnoreCondition`, it follows from a consistency standpoint that the same should apply to `IgnoreReadOnlyProperties` and `IgnoreReadOnlyFields`. However, this is what this PR is trying to change, since global options configuration ends up overriding a user-provided `Set` delegate.
2. Ultimately, this is about us deciding what the semantics of `Options.DefaultIgnoreCondition` are vis-a-vis `IJsonTypeInfoResolver`: it could be an extraneous setting applied indiscriminately by the serializer on generated contracts _or_ a parameter being consulted by `IJsonTypeInfoResolver` (which it already does for a bunch of other knobs). Neither approach has any inherent inconsistencies, so this is merely an issue of picking a convention. Because of 1) I maintain we should go with the latter convention.

Reply:
>  it follows from a consistency standpoint that the same should apply to IgnoreReadOnlyProperties and IgnoreReadOnlyFields

I really disagree on that. `null` value in most of the framework refers to `not sepcified` or `default`. For `ShouldSerialize` first impression is that `null` should mean options default. For `IgnoreReadOnlyProperties` and `IgnoreReadOnlyFields` situation is different since name is not declarative like `DefaultIgnoreCondition`. It purely suggest we should be ignoring read only properties and it would seem really weird if we ignored field or properties after user has explicitly set metadata especially at the point where `property` and `field` are meaningless. The only place where it seems it could be used is for generating metadata and that is not confusing IMO.

I'd personally find it much more confusing if `null` meant something else than `default` in this context.

Reply:
btw. this PR is doing much more than just what title is suggesting and I'd suggest to scope the fix here to what title says and remaining part we can discuss separately in an issue

Reply:
The PR is doing precisely what is enumerated in the description. Let me state clearly that the changes are not unrelated. Please be cognizant of the time constraints, I don't believe breaking this up now would be productive. Let's address the concern and hand and get this fix merged.

Reply:
I've reverted the particular change. I still don't agree that it should have been reverted, but we need to get this fix in.